# Data understanding


In [473]:
# importing libraries
import pandas as pd
import sqlite3

In [474]:
# loading data sets
bom_df= pd.read_csv("data/bom.movie_gross.csv")
rt_reviews_df=pd.read_csv("data/rt.reviews.tsv", sep= '\t', encoding= 'latin1') 
rt_movie_info_df= pd.read_csv("data/rt.movie_info.tsv", sep= '\t', encoding= 'latin1')
tmdb_movies_df= pd.read_csv("data/tmdb.movies.csv")
tn_movie_budgets_df= pd.read_csv("data/tn.movie_budgets.csv")

# previewing each
for name, df in [('Box office mojo', bom_df), ('Rotten tomatoes Reviews', rt_reviews_df), ('Rotten tomatoes movie info', rt_movie_info_df),
                 ('The movie DB',tmdb_movies_df), ('The numbers', tn_movie_budgets_df)]:
    print(f"\n{name} Sample:")
    print(df.head())
    print(df.info())



Box office mojo Sample:
                                         title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year       

## Data understanding of each dataframe

In [475]:
#Box office mojo dataset(movie_gross)
# checking the shape
bom_df= pd.read_csv("data/bom.movie_gross.csv")
bom_df.shape

(3387, 5)

bom_df dataframe from Box office Mojo has 3387 records and 5 columns.

In [476]:
# checking the info
bom_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


## Observation

Total rows: 3387 movies

Columns: 5 - title, studio, domestic_gross, foreign_gross and year.

The dataset is mostly complete with title and year having no missing values

The column with more missing values is foreign_gross followed by domestic_gross and lastly studio.

foreign_gross is stored as object instead of float, likely due to formatting issues.


In [477]:
# reviewing Rotten Tomatoes dataset(reviews df)
rt_reviews_df=pd.read_csv("data/rt.reviews.tsv", sep= '\t', encoding= 'latin1') 
rt_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54432 entries, 0 to 54431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          54432 non-null  int64 
 1   review      48869 non-null  object
 2   rating      40915 non-null  object
 3   fresh       54432 non-null  object
 4   critic      51710 non-null  object
 5   top_critic  54432 non-null  int64 
 6   publisher   54123 non-null  object
 7   date        54432 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.3+ MB


In [478]:
# checking missing values
rt_reviews_df.isna().sum()

id                0
review         5563
rating        13517
fresh             0
critic         2722
top_critic        0
publisher       309
date              0
dtype: int64

## Observation

Total rows: 54,432 reviews

Columns: 8 including id, review, rating, fresh, critic, top_critic, publisher, and date

All rows have id, fresh, top_critic, and date

missing values are significant in some columns

- review: 5563 missing values

- rating: 13517 missing

- rating: 13517 missing

- publisher: 309 missing values

Date column can be converted to date-time format



In [479]:
# reviewing Rotten Tomatoes dataset(movie_info)
rt_movie_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1560 entries, 0 to 1559
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            1560 non-null   int64 
 1   synopsis      1498 non-null   object
 2   rating        1557 non-null   object
 3   genre         1552 non-null   object
 4   director      1361 non-null   object
 5   writer        1111 non-null   object
 6   theater_date  1201 non-null   object
 7   dvd_date      1201 non-null   object
 8   currency      340 non-null    object
 9   box_office    340 non-null    object
 10  runtime       1530 non-null   object
 11  studio        494 non-null    object
dtypes: int64(1), object(11)
memory usage: 146.4+ KB


In [480]:
rt_movie_info_df.isna().sum()

id                 0
synopsis          62
rating             3
genre              8
director         199
writer           449
theater_date     359
dvd_date         359
currency        1220
box_office      1220
runtime           30
studio          1066
dtype: int64

## Observation
Total rows: 1560 movie reviews
    
Columns: 12- id, synopsis, rating, genre, director, writer, theater_date, dvd_date, currency, box_office, runtime, studio

Columns like id, rating, genre, and runtime are mostly complete.

Missing values are significant in several columns

- synopsis: 62 missing

- director: 199 missing

- writer: 449 missing

- theater_date and dvd_date: both have 359 missing

- box_office and currency: 1220 missing

- studio: 1066 missing, most data missing

Theater_date, dvd_date, and runtime should be converted to datetime or numeric for analysis.

box_office and currency are object types need to be converted to numeric




In [481]:
#  reviewing the movie DB dataset(tmdb_movie df)
tmdb_movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26517 entries, 0 to 26516
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         26517 non-null  int64  
 1   genre_ids          26517 non-null  object 
 2   id                 26517 non-null  int64  
 3   original_language  26517 non-null  object 
 4   original_title     26517 non-null  object 
 5   popularity         26517 non-null  float64
 6   release_date       26517 non-null  object 
 7   title              26517 non-null  object 
 8   vote_average       26517 non-null  float64
 9   vote_count         26517 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 2.0+ MB


## Observation

Total movies: 26,517

Columns: 10 - includes title info, ratings, language, and release_date
    
No missing values across any of the 10 columns - very clean.

unnamed column needs to be dropped

release_date should be converted to datetime

In [482]:
# reviewing The Numbers dataset(movie_budgets df)
tn_movie_budgets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB


## Observation

Total entries: 5,782 movies

Columns: 6 - focused on financial and release information

Complete data (no missing entries) across all columns

Key financial metrics included:

- production_budget

- domestic_gross

- worldwide_gross

release_date is present, enabling time-based trend analysis

In [483]:
# loading the sqlite database
con= sqlite3.connect("data/im.db") #creating a connection

# previewing the schema
pd.read_sql("""
SELECT name
FROM sqlite_master
WHERE type= "table";
""", con)

,name
0,movie_basics
1,directors
2,known_for
3,movie_akas
4,movie_ratings
5,persons
6,principals
7,writers


In [484]:
# selecting movie_basics and movie_ratings tables
query= """
SELECT *
FROM movie_basics

"""
pd.read_sql(query, con).tail()

,movie_id,primary_title,original_title,start_year,runtime_minutes,genres
146139,tt9916538,Kuambil Lagi Hatiku,Kuambil Lagi Hatiku,2019,123.0,Drama
146140,tt9916622,Rodolpho Teóphilo - O Legado de um Pioneiro,Rodolpho Teóphilo - O Legado de um Pioneiro,2015,NaN,Documentary
146141,tt9916706,Dankyavar Danka,Dankyavar Danka,2013,NaN,Comedy
146142,tt9916730,6 Gunn,6 Gunn,2017,116.0,None
146143,tt9916754,Chico Albuquerque - Revelações,Chico Albuquerque - Revelações,2013,NaN,Documentary


In [485]:
query= """
SELECT *
FROM movie_ratings
"""
pd.read_sql(query, con).tail()

,movie_id,averagerating,numvotes
73851,tt9805820,8.1,25
73852,tt9844256,7.5,24
73853,tt9851050,4.7,14
73854,tt9886934,7.0,5
73855,tt9894098,6.3,128


In [486]:
# joining the two tables using movie_id
query= """
SELECT *
FROM movie_basics
LEFT JOIN movie_ratings
USING(movie_id);
"""
movie_basics_rating_df= pd.read_sql(query, con)
movie_basics_rating_df.tail()

,movie_id,primary_title,original_title,start_year,runtime_minutes,genres,averagerating,numvotes
146139,tt9916538,Kuambil Lagi Hatiku,Kuambil Lagi Hatiku,2019,123.0,Drama,NaN,NaN
146140,tt9916622,Rodolpho Teóphilo - O Legado de um Pioneiro,Rodolpho Teóphilo - O Legado de um Pioneiro,2015,NaN,Documentary,NaN,NaN
146141,tt9916706,Dankyavar Danka,Dankyavar Danka,2013,NaN,Comedy,NaN,NaN
146142,tt9916730,6 Gunn,6 Gunn,2017,116.0,None,NaN,NaN
146143,tt9916754,Chico Albuquerque - Revelações,Chico Albuquerque - Revelações,2013,NaN,Documentary,NaN,NaN


# Data cleaning

## Box office mojo dataset

In [487]:
# creating a working copy 
bom_clean_df= bom_df.copy()

In [488]:
bom_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


In [489]:
# checking for duplicates
duplicated_rows= bom_clean_df.duplicated()
duplicated_rows.sum()

0

In [490]:
# cleaning the bom_df(movie_gross) from Box office mojo dataset
bom_clean_df.columns

Index(['title', 'studio', 'domestic_gross', 'foreign_gross', 'year'], dtype='object')

In [491]:
# cleaning title column
# checking for duplicates
key_columns= ['title', 'year']
dupl_row= bom_clean_df[bom_clean_df.duplicated(subset= key_columns, keep= False)]
dupl_row

,title,studio,domestic_gross,foreign_gross,year


There are no duplicates

In [492]:
bom_clean_df.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

In [493]:
# cleaning studio column
bom_clean_df['studio'] = bom_clean_df['studio'].str.strip()

In [494]:
# checking for missing values 
missing_studio = bom_clean_df['studio'].isnull().sum()
missing_studio

# filling in missing values
bom_clean_df['studio']= bom_clean_df['studio'].fillna("unknown")

In [495]:
# checking for missing values
bom_clean_df['studio'].isnull().sum()

0

In [496]:
missing_domestic_gross = bom_clean_df[bom_clean_df['domestic_gross'].isna()]
# Print the rows with missing domestic_gross
print("\nRows with missing domestic_gross:")
print(missing_domestic_gross[['title', 'studio', 'domestic_gross', 'foreign_gross', 'year']])


Rows with missing domestic_gross:
                                      title   studio  domestic_gross  \
230              It's a Wonderful Afterlife      UTV             NaN   
298   Celine: Through the Eyes of the World     Sony             NaN   
302                              White Lion    Scre.             NaN   
306                        Badmaash Company     Yash             NaN   
327                      Aashayein (Wishes)  Relbig.             NaN   
537                                   Force     FoxS             NaN   
713                        Empire of Silver     NeoC             NaN   
871                            Solomon Kane     RTWC             NaN   
928                            The Tall Man    Imag.             NaN   
933                   Keith Lemon: The Film  unknown             NaN   
936                     Lula, Son of Brazil     NYer             NaN   
966                          The Cup (2012)     Myr.             NaN   
1017                         

In [497]:
# filling in missing values with mean
mean= bom_clean_df['domestic_gross'].mean()
bom_clean_df['domestic_gross']= bom_clean_df['domestic_gross'].fillna(mean)


# checking for null
bom_clean_df['domestic_gross'].isnull().sum()



0

In [498]:
# cleaning year column
bom_clean_df["year"].dtype

dtype('int64')

In [499]:
# cleaning foreign column
bom_clean_df['foreign_gross'].isnull().sum()


1350

In [500]:
# dropping foreign_gross column
bom_clean_df.drop(columns= 'foreign_gross', inplace= True)

## Cleaning Rotten Tomatoes reviews dataset


In [501]:
# creating a copy
rt_reviews_clean_df=rt_reviews_df.copy()

In [502]:
# missing values
rt_reviews_clean_df.isna().sum()

id                0
review         5563
rating        13517
fresh             0
critic         2722
top_critic        0
publisher       309
date              0
dtype: int64

In [503]:
# checking for id duplicates
duplicate_count = rt_reviews_clean_df.duplicated().sum()
print(f"Number of duplicates: {duplicate_count}")

Number of duplicates: 9


In [504]:
# checking for duplicates while focusing on specific columns for clarity
if all(col in rt_reviews_clean_df for col in ['id', 'review', 'critic','publisher', 'date']):
        initial_rows_rt_reviews = len(rt_reviews_clean_df)
        rt_reviews_clean_df.drop_duplicates(subset=['id', 'review', 'critic','publisher', 'date'])
        print(f"rt_reviews_clean_df: Removed {initial_rows_rt_reviews - len(rt_reviews_clean_df)} duplicates based on 'id', 'review','publisher', 'critic', and 'date'.")
else:
        print("rt_reviews_clean_df: Skipping duplicate check due to missing 'id', 'review', 'critic', or 'date' columns.")

rt_reviews_clean_df: Removed 0 duplicates based on 'id', 'review','publisher', 'critic', and 'date'.


In [505]:
# checking for columns unique values in each column
for column in rt_reviews_clean_df.columns:
    print(f"\n{column} unique values:")
    print(rt_reviews_df[column].unique())


id unique values:
[   3    5    6 ... 1998 1999 2000]

review unique values:
["A distinctly gallows take on contemporary financial mores, as one absurdly rich man's limo ride across town for a haircut functions as a state-of-the-nation discourse. "
 "It's an allegory in search of a meaning that never arrives...It's just old-fashioned bad storytelling."
 '... life lived in a bubble in financial dealings and digital communications and brief face-to-face conversations and sexual intermissions in a space shuttle of a limousine creeping through the gridlock of an anonymous New York City.'
 ...
 "Despite Besson's high-profile name being Wasabi's big selling point, there is no doubt that Krawczyk deserves a huge amount of the credit for the film's thoroughly winning tone."
 'The film lapses too often into sugary sentiment and withholds delivery on the pell-mell pyrotechnics its punchy style promises.'
 'The real charm of this trifle is the deadpan comic face of its star, Jean Reno, who resem

In [506]:
# filling in review column
if 'review' in rt_reviews_clean_df.columns:
    missing_review_count= rt_reviews_clean_df['review'].isna().sum()
    if missing_review_count >0:
        rt_reviews_clean_df['review']= rt_reviews_clean_df['review'].fillna('No Review Text')
        print(f"Filled {missing_review_count} missing 'review' values with 'No Review Text'.")

Filled 5563 missing 'review' values with 'No Review Text'.


In [507]:
# dropping unuseful columns
rt_reviews_clean_df.drop(columns=['rating'], inplace=True)

In [508]:
# checking if the columns are dropped
rt_reviews_clean_df.columns

Index(['id', 'review', 'fresh', 'critic', 'top_critic', 'publisher', 'date'], dtype='object')

In [509]:
# filling critic column missing values with unknown
rt_reviews_clean_df['critic']= rt_reviews_clean_df['critic'].fillna("Unknown")

In [510]:
# filling publisher missing with unknown
rt_reviews_clean_df['publisher']= rt_reviews_clean_df['publisher'].fillna("Unknown")

In [511]:
# converting date into datetime format
rt_reviews_clean_df['date'] = pd.to_datetime(rt_reviews_clean_df['date'])

# checking effectiveness
rt_reviews_clean_df["date"].dtype

dtype('<M8[ns]')

In [512]:
# checking missing values in the df
rt_reviews_clean_df.isna().sum()

id            0
review        0
fresh         0
critic        0
top_critic    0
publisher     0
date          0
dtype: int64

## Cleaning Rotten Tomatoes movie info dataset

In [513]:
# creating a working copy
rt_movie_info_clean= rt_movie_info_df.copy()


In [514]:
# checking for missing values
rt_movie_info_clean.isnull().sum()

id                 0
synopsis          62
rating             3
genre              8
director         199
writer           449
theater_date     359
dvd_date         359
currency        1220
box_office      1220
runtime           30
studio          1066
dtype: int64

In [515]:
# checking for columns unique values in each column
for column in rt_movie_info_clean.columns:
    print(f"\n{column} unique values:")
    print(rt_movie_info_clean[column].unique())


id unique values:
[   1    3    5 ... 1998 1999 2000]

synopsis unique values:
['This gritty, fast-paced, and innovative police drama earned five Academy Awards, including Best Picture, Best Adapted Screenplay (written by Ernest Tidyman), and Best Actor (Gene Hackman). Jimmy "Popeye" Doyle (Hackman) and his partner, Buddy Russo (Roy Scheider), are New York City police detectives on narcotics detail, trying to track down the source of heroin from Europe into the United States. Suave Alain Charnier (Fernando Rey) is the French drug kingpin who provides a large percentage of New York City\'s dope, and Pierre Nicoli (Marcel Bozzuffi) is a hired killer and Charnier\'s right-hand man. Acting on a hunch, Popeye and Buddy start tailing Sal Boca (Tony Lo Bianco) and his wife, Angie (Arlene Faber), who live pretty high for a couple whose corner store brings in about 7,000 dollars a year. It turns out Popeye\'s suspicions are right -- Sal and Angie are the New York agents for Charnier, who will 

## Handling missing values column by column


### Filling in categorical columns with 'unknown'


In [516]:

for col in ['synopsis', 'rating', 'genre', 'director', 'writer', 'studio']:
    if col in rt_movie_info_clean.columns:
            initial_missing_count = rt_movie_info_clean[col].isnull().sum()
            if initial_missing_count > 0:
                rt_movie_info_clean[col] = rt_movie_info_clean[col].fillna('Unknown')
                print(f"Filled {initial_missing_count} missing '{col}' values with 'Unknown'.")


Filled 62 missing 'synopsis' values with 'Unknown'.
Filled 3 missing 'rating' values with 'Unknown'.
Filled 8 missing 'genre' values with 'Unknown'.
Filled 199 missing 'director' values with 'Unknown'.
Filled 449 missing 'writer' values with 'Unknown'.
Filled 1066 missing 'studio' values with 'Unknown'.


In [517]:
rt_movie_info_clean.isna().sum()

id                 0
synopsis           0
rating             0
genre              0
director           0
writer             0
theater_date     359
dvd_date         359
currency        1220
box_office      1220
runtime           30
studio             0
dtype: int64

In [518]:
# converting theatre date and dvd date into datetime format
rt_movie_info_clean['theater_date'] = pd.to_datetime(rt_movie_info_clean['theater_date'], errors='coerce')
rt_movie_info_clean['dvd_date'] = pd.to_datetime(rt_movie_info_clean['dvd_date'], errors='coerce')

In [519]:
# filling in missing runtime values with median runtime
rt_movie_info_clean["runtime"]= rt_movie_info_clean["runtime"].astype(str).str.replace('minutes', '', regex= False)
rt_movie_info_clean["runtime"]= pd.to_numeric(rt_movie_info_clean["runtime"], errors= 'coerce')
# filling missing values with median values
runtime_median= rt_movie_info_clean["runtime"].median()                                     
rt_movie_info_clean["runtime"]= rt_movie_info_clean["runtime"].fillna(runtime_median)

In [520]:
rt_movie_info_clean.isna().sum()

id                 0
synopsis           0
rating             0
genre              0
director           0
writer             0
theater_date     359
dvd_date         359
currency        1220
box_office      1220
runtime            0
studio             0
dtype: int64

In [521]:
# dropping missing theater date and dvd date missing values
rt_movie_info_clean = rt_movie_info_clean.drop(columns=['currency', 'box_office'])


## Cleaning tmdb_movies_df(The movie DB dataset)

In [522]:
tmdb_movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26517 entries, 0 to 26516
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         26517 non-null  int64  
 1   genre_ids          26517 non-null  object 
 2   id                 26517 non-null  int64  
 3   original_language  26517 non-null  object 
 4   original_title     26517 non-null  object 
 5   popularity         26517 non-null  float64
 6   release_date       26517 non-null  object 
 7   title              26517 non-null  object 
 8   vote_average       26517 non-null  float64
 9   vote_count         26517 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 2.0+ MB


In [523]:
# dropping unnamed column
tmdb_movies_df.drop(columns=["Unnamed: 0", 'genre_ids'], inplace=True)

In [524]:
# checking for unique values in each column
for column in tmdb_movies_df.columns:
    print(f"\n{column} unique values:")
    print(tmdb_movies_df[column].unique())


id unique values:
[ 12444  10191  10138 ... 381231 366854 309885]

original_language unique values:
['en' 'nl' 'es' 'ja' 'sv' 'de' 'fr' 'cn' 'it' 'ru' 'zh' 'hi' 'no' 'ko'
 'da' 'fi' 'pl' 'te' 'hu' 'tr' 'pt' 'he' 'fa' 'th' 'cs' 'et' 'tl' 'lt'
 'xx' 'bs' 'ar' 'is' 'el' 'mr' 'hr' 'ro' 'sr' 'uk' 'nb' 'hz' 'ca' 'bg'
 'sl' 'lv' 'si' 'ab' 'ta' 'bo' 'id' 'sq' 'bn' 'gu' 'lo' 'ne' 'kk' 'hy'
 'ps' 'kn' 'vi' 'ku' 'ka' 'ml' 'ur' 'mi' 'eu' 'sn' 'cy' 'ha' 'ky' 'yi'
 'pa' 'xh' 'cr' 'sw' 'af' 'dz']

original_title unique values:
['Harry Potter and the Deathly Hallows: Part 1' 'How to Train Your Dragon'
 'Iron Man 2' ... 'The Last One' 'Trailer Made' 'The Church']

popularity unique values:
[33.533 28.734 28.515 ...  0.612  0.609  0.607]

release_date unique values:
['2010-11-19' '2010-03-26' '2010-05-07' ... '2018-05-08' '2018-08-02'
 '2018-05-26']

title unique values:
['Harry Potter and the Deathly Hallows: Part 1' 'How to Train Your Dragon'
 'Iron Man 2' ... 'The Last One' 'Trailer Made' 'The Churc

In [525]:
# standardizing categorical columns for consistency (lowercase and strip)
tmdb_movies_df['original_language']= tmdb_movies_df['original_language'].str.strip().str.lower()
tmdb_movies_df['original_title']= tmdb_movies_df['original_title'].str.strip()
tmdb_movies_df['title']= tmdb_movies_df['title'].str.strip()

In [526]:
# converting release date to datetime format
tmdb_movies_df["release_date"] = pd.to_datetime(tmdb_movies_df["release_date"], errors='coerce')

In [527]:
# extract year and month from movie_budgets tmdb df
# tmdb_movies_df['release_year'] = tmdb_movies_df['release_date'].dt.year
# tmdb_movies_df['release_month'] = tmdb_movies_df['release_date'].dt.month

In [528]:
tmdb_movies_df.columns

Index(['id', 'original_language', 'original_title', 'popularity',
       'release_date', 'title', 'vote_average', 'vote_count'],
      dtype='object')

## Cleaning movie_budgets df(The Numbers Dataset)

In [529]:
tn_movie_budgets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB


In [530]:
# checking for unique values in each column
for column in tn_movie_budgets_df.columns:
    print(f"\n{column} unique values:")
    print(tn_movie_budgets_df[column].unique())


id unique values:
[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100]

release_date unique values:
['Dec 18, 2009' 'May 20, 2011' 'Jun 7, 2019' ... 'Aug 14, 2008'
 'Feb 26, 1993' 'Jul 13, 2005']

movie unique values:
['Avatar' 'Pirates of the Caribbean: On Stranger Tides' 'Dark Phoenix' ...
 'Return to the Land of Wonders' 'A Plague So Pleasant'
 'My Date With Drew']

production_budget unique values:
['$425,000,000' '$410,600,000' '$350,000,000' '$330,600,000'
 '$317,000,000' '$306,000,000' '$300,000,000' '$275,000,000'
 '$260,000,000' '$258,000,000' '$250,000,000' '$232,000,000'
 '$230,000,000' '$225,000,000' '$220,000,000' '$217,000,0

In [531]:
# converting currency strings to numeric 
print("\ntn_movie_budgets_df...")
for col in ['production_budget', 'domestic_gross', 'worldwide_gross']:
        if col in tn_movie_budgets_df.columns and tn_movie_budgets_df[col].dtype == 'object':
            tn_movie_budgets_df[col] = tn_movie_budgets_df[col].astype(str).str.replace('$', '', regex=False).str.replace(',', '', regex=False)
            tn_movie_budgets_df[col] = pd.to_numeric(tn_movie_budgets_df[col], errors='coerce')


tn_movie_budgets_df...


In [532]:
# checking 
tn_movie_budgets_df.head(10)

,id,release_date,movie,production_budget,domestic_gross,worldwide_gross
0,1,"Dec 18, 2009",Avatar,425000000,760507625,2776345279
1,2,"May 20, 2011",Pirates of the Caribbean: On Stranger Tides,410600000,241063875,1045663875
2,3,"Jun 7, 2019",Dark Phoenix,350000000,42762350,149762350
3,4,"May 1, 2015",Avengers: Age of Ultron,330600000,459005868,1403013963
4,5,"Dec 15, 2017",Star Wars Ep. VIII: The Last Jedi,317000000,620181382,1316721747
5,6,"Dec 18, 2015",Star Wars Ep. VII: The Force Awakens,306000000,936662225,2053311220
6,7,"Apr 27, 2018",Avengers: Infinity War,300000000,678815482,2048134200
7,8,"May 24, 2007",Pirates of the Caribbean: At World’s End,300000000,309420425,963420425
8,9,"Nov 17, 2017",Justice League,300000000,229024295,655945209
9,10,"Nov 6, 2015",Spectre,300000000,200074175,879620923


In [533]:
# converting release date to datetime format
tn_movie_budgets_df['release_date']= pd.to_datetime(tn_movie_budgets_df['release_date'])

In [534]:
# Extract release year and month from tn budget
tn_movie_budgets_df['release_year'] = tn_movie_budgets_df['release_date'].dt.year
tn_movie_budgets_df['release_month'] = tn_movie_budgets_df['release_date'].dt.month


In [535]:
tn_movie_budgets_df.columns

Index(['id', 'release_date', 'movie', 'production_budget', 'domestic_gross',
       'worldwide_gross', 'release_year', 'release_month'],
      dtype='object')

## Feature engineering

### Calculating review counts from rt_reviews_clean_df

In [536]:
# computing review counts and freshness score
review_counts= rt_reviews_clean_df.groupby('id').agg(
    Review_counts= ('fresh', 'count'),
    Freshness_score= ('fresh', lambda x: (x== 'fresh').sum()/ len(x))).reset_index()
review_counts

,id,Review_counts,Freshness_score
0,3,163,0.631902
1,5,23,0.782609
2,6,57,0.561404
3,8,75,0.746667
4,10,108,0.462963
...,...,...,...
1130,1996,143,0.671329
1131,1997,28,0.357143
1132,1998,2,1.000000
1133,1999,46,0.586957


### Merging review_counts and rt_movie_info_clean df

In [537]:
rt_combined_df= pd.merge(rt_movie_info_clean, review_counts, on= 'id', how= 'left')
rt_combined_df

,id,synopsis,rating,genre,director,writer,theater_date,dvd_date,runtime,studio,Review_counts,Freshness_score
0,1,"This gritty, fast-paced, and innovative police...",R,Action and Adventure|Classics|Drama,William Friedkin,Ernest Tidyman,1971-10-09,2001-09-25,104.0,Unknown,NaN,NaN
1,3,"New York City, not-too-distant-future: Eric Pa...",R,Drama|Science Fiction and Fantasy,David Cronenberg,David Cronenberg|Don DeLillo,2012-08-17,2013-01-01,108.0,Entertainment One,163.0,0.631902
2,5,Illeana Douglas delivers a superb performance ...,R,Drama|Musical and Performing Arts,Allison Anders,Allison Anders,1996-09-13,2000-04-18,116.0,Unknown,23.0,0.782609
3,6,Michael Douglas runs afoul of a treacherous su...,R,Drama|Mystery and Suspense,Barry Levinson,Paul Attanasio|Michael Crichton,1994-12-09,1997-08-27,128.0,Unknown,57.0,0.561404
4,7,Unknown,NR,Drama|Romance,Rodney Bennett,Giles Cooper,NaT,NaT,200.0,Unknown,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
1555,1996,Forget terrorists or hijackers -- there's a ha...,R,Action and Adventure|Horror|Mystery and Suspense,Unknown,Unknown,2006-08-18,2007-01-02,106.0,New Line Cinema,143.0,0.671329
1556,1997,The popular Saturday Night Live sketch was exp...,PG,Comedy|Science Fiction and Fantasy,Steve Barron,Terry Turner|Tom Davis|Dan Aykroyd|Bonnie Turner,1993-07-23,2001-04-17,88.0,Paramount Vantage,28.0,0.357143
1557,1998,"Based on a novel by Richard Powell, when the l...",G,Classics|Comedy|Drama|Musical and Performing Arts,Gordon Douglas,Unknown,1962-01-01,2004-05-11,111.0,Unknown,2.0,1.000000
1558,1999,The Sandlot is a coming-of-age story about a g...,PG,Comedy|Drama|Kids and Family|Sports and Fitness,David Mickey Evans,David Mickey Evans|Robert Gunter,1993-04-01,2002-01-29,101.0,Unknown,46.0,0.586957


### Merging tmdb_movies_df and tn_movie_budgets_df


In [538]:
# renaming tmdb title column to tmdb_title for differentiation
tmdb_movies_df.rename(columns= {'title': 'tmdb_title'}, inplace= True)

In [539]:
# renaming movie column in tn_movie_budgets_df for easy understanding when merging with tmdb
tn_movie_budgets_df.rename(columns={ 'movie': 'title'}, inplace= True)

In [543]:
# merging tn_movie_budgets_df and tmdb_movies_df on title and release year
tmdb_tn_merged= pd.merge(tn_movie_budgets_df, tmdb_movies_df, left_on= ['title', 'release_year'], right_on=['tmdb_title', 'release_year'], how= 'left')
tmdb_tn_merged

,id_x,release_date_x,title,production_budget,domestic_gross,worldwide_gross,release_year,release_month_x,id_y,original_language,original_title,popularity,release_date_y,tmdb_title,vote_average,vote_count,release_month_y
0,1,2009-12-18,Avatar,425000000,760507625,2776345279,2009,12,19995.0,en,Avatar,26.526,2009-12-18,Avatar,7.4,18676.0,12.0
1,2,2011-05-20,Pirates of the Caribbean: On Stranger Tides,410600000,241063875,1045663875,2011,5,1865.0,en,Pirates of the Caribbean: On Stranger Tides,30.579,2011-05-20,Pirates of the Caribbean: On Stranger Tides,6.4,8571.0,5.0
2,3,2019-06-07,Dark Phoenix,350000000,42762350,149762350,2019,6,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
3,4,2015-05-01,Avengers: Age of Ultron,330600000,459005868,1403013963,2015,5,99861.0,en,Avengers: Age of Ultron,44.383,2015-05-01,Avengers: Age of Ultron,7.3,13457.0,5.0
4,5,2017-12-15,Star Wars Ep. VIII: The Last Jedi,317000000,620181382,1316721747,2017,12,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5935,78,2018-12-31,Red 11,7000,0,0,2018,12,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
5936,79,1999-04-02,Following,6000,48482,240495,1999,4,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
5937,80,2005-07-13,Return to the Land of Wonders,5000,1338,1338,2005,7,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
5938,81,2015-09-29,A Plague So Pleasant,1400,0,0,2015,9,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN


In [544]:
# Rename id_x for merging
tmdb_tn_merged.rename(columns={'id_x': 'id'}, inplace=True)

In [547]:
merged_2= pd.merge(tmdb_tn_merged, rt_combined_df[['id', 'genre', 'rating', 'Review_counts', 'Freshness_score']], on='id', how='left')
merged_2

,id,release_date_x,title,production_budget,domestic_gross,worldwide_gross,release_year,release_month_x,id_y,original_language,...,popularity,release_date_y,tmdb_title,vote_average,vote_count,release_month_y,genre,rating,Review_counts,Freshness_score
0,1,2009-12-18,Avatar,425000000,760507625,2776345279,2009,12,19995.0,en,...,26.526,2009-12-18,Avatar,7.4,18676.0,12.0,Action and Adventure|Classics|Drama,R,NaN,NaN
1,2,2011-05-20,Pirates of the Caribbean: On Stranger Tides,410600000,241063875,1045663875,2011,5,1865.0,en,...,30.579,2011-05-20,Pirates of the Caribbean: On Stranger Tides,6.4,8571.0,5.0,NaN,NaN,NaN,NaN
2,3,2019-06-07,Dark Phoenix,350000000,42762350,149762350,2019,6,NaN,NaN,...,NaN,NaT,NaN,NaN,NaN,NaN,Drama|Science Fiction and Fantasy,R,163.0,0.631902
3,4,2015-05-01,Avengers: Age of Ultron,330600000,459005868,1403013963,2015,5,99861.0,en,...,44.383,2015-05-01,Avengers: Age of Ultron,7.3,13457.0,5.0,NaN,NaN,NaN,NaN
4,5,2017-12-15,Star Wars Ep. VIII: The Last Jedi,317000000,620181382,1316721747,2017,12,NaN,NaN,...,NaN,NaT,NaN,NaN,NaN,NaN,Drama|Musical and Performing Arts,R,23.0,0.782609
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5935,78,2018-12-31,Red 11,7000,0,0,2018,12,NaN,NaN,...,NaN,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5936,79,1999-04-02,Following,6000,48482,240495,1999,4,NaN,NaN,...,NaN,NaT,NaN,NaN,NaN,NaN,Action and Adventure|Drama|Mystery and Suspense,R,18.0,0.611111
5937,80,2005-07-13,Return to the Land of Wonders,5000,1338,1338,2005,7,NaN,NaN,...,NaN,NaT,NaN,NaN,NaN,NaN,Action and Adventure|Art House and Internation...,NR,15.0,0.866667
5938,81,2015-09-29,A Plague So Pleasant,1400,0,0,2015,9,NaN,NaN,...,NaN,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Final feature engineering

In [550]:
merged_2['Profit']= merged_2['worldwide_gross']- merged_2['production_budget']
merged_2['ROI'] = merged_2['Profit'] / merged_2['production_budget']
merged_2['Domestic_vs_Worldwide_Ratio'] = merged_2['domestic_gross'] / merged_2['worldwide_gross']
merged_2['Release_Month'] = merged_2['release_date_x'].dt.month
merged_2['Day_of_Week'] = merged_2['release_date_x'].dt.day_name()


In [551]:
merged_2.head(10)

,id,release_date_x,title,production_budget,domestic_gross,worldwide_gross,release_year,release_month_x,id_y,original_language,...,genre,rating,Review_counts,Freshness_score,profit,Profit,ROI,Domestic_vs_Worldwide_Ratio,Release_Month,Day_of_Week
0,1,2009-12-18,Avatar,425000000,760507625,2776345279,2009,12,19995.0,en,...,Action and Adventure|Classics|Drama,R,NaN,NaN,2351345279,2351345279,5.532577,0.273924,12,Friday
1,2,2011-05-20,Pirates of the Caribbean: On Stranger Tides,410600000,241063875,1045663875,2011,5,1865.0,en,...,NaN,NaN,NaN,NaN,635063875,635063875,1.546673,0.230537,5,Friday
2,3,2019-06-07,Dark Phoenix,350000000,42762350,149762350,2019,6,NaN,NaN,...,Drama|Science Fiction and Fantasy,R,163.0,0.631902,-200237650,-200237650,-0.572108,0.285535,6,Friday
3,4,2015-05-01,Avengers: Age of Ultron,330600000,459005868,1403013963,2015,5,99861.0,en,...,NaN,NaN,NaN,NaN,1072413963,1072413963,3.243841,0.327157,5,Friday
4,5,2017-12-15,Star Wars Ep. VIII: The Last Jedi,317000000,620181382,1316721747,2017,12,NaN,NaN,...,Drama|Musical and Performing Arts,R,23.0,0.782609,999721747,999721747,3.153696,0.471004,12,Friday
5,6,2015-12-18,Star Wars Ep. VII: The Force Awakens,306000000,936662225,2053311220,2015,12,NaN,NaN,...,Drama|Mystery and Suspense,R,57.0,0.561404,1747311220,1747311220,5.710167,0.456172,12,Friday
6,7,2018-04-27,Avengers: Infinity War,300000000,678815482,2048134200,2018,4,299536.0,en,...,Drama|Romance,NR,NaN,NaN,1748134200,1748134200,5.827114,0.331431,4,Friday
7,8,2007-05-24,Pirates of the Caribbean: At World’s End,300000000,309420425,963420425,2007,5,NaN,NaN,...,Drama|Kids and Family,PG,75.0,0.746667,663420425,663420425,2.211401,0.321169,5,Thursday
8,9,2017-11-17,Justice League,300000000,229024295,655945209,2017,11,141052.0,en,...,NaN,NaN,NaN,NaN,355945209,355945209,1.186484,0.349152,11,Friday
9,9,2017-11-17,Justice League,300000000,229024295,655945209,2017,11,141052.0,en,...,NaN,NaN,NaN,NaN,355945209,355945209,1.186484,0.349152,11,Friday


In [552]:
merged_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5940 entries, 0 to 5939
Data columns (total 27 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   id                           5940 non-null   int64         
 1   release_date_x               5940 non-null   datetime64[ns]
 2   title                        5940 non-null   object        
 3   production_budget            5940 non-null   int64         
 4   domestic_gross               5940 non-null   int64         
 5   worldwide_gross              5940 non-null   int64         
 6   release_year                 5940 non-null   int64         
 7   release_month_x              5940 non-null   int64         
 8   id_y                         1758 non-null   float64       
 9   original_language            1758 non-null   object        
 10  original_title               1758 non-null   object        
 11  popularity                   1758 non-null 

In [557]:
# Drop suffix columns and low-value metadata
redundant_cols = [col for col in merged_2.columns if col.endswith('_x') or col.endswith('_y')]
low_value_cols = [
    'id_y', 'tmdb_title', 'original_title', 'original_language', 'currency',
    'dvd_date', 'release_date_y', 'release_date_x', 'rating', 'writer',
    'synopsis', 'theater_date', 'popularity', 'vote_count', 'vote_average','Unnamed: 0','genre_ids'
]
drop_columns = list(set(redundant_cols + low_value_cols))
merged_2.drop(columns=drop_columns, errors='ignore', inplace=True)
merged_2.tail()

,id,title,production_budget,domestic_gross,worldwide_gross,release_year,genre,Review_counts,Freshness_score,profit,Profit,ROI,Domestic_vs_Worldwide_Ratio,Release_Month,Day_of_Week
5934,77,The Mongol King,7000,900,900,2004,Comedy|Drama,183.0,0.360656,-6100,-6100,-0.871429,1.0,12,Friday
5935,78,Red 11,7000,0,0,2018,NaN,NaN,NaN,-7000,-7000,-1.000000,NaN,12,Monday
5937,80,Return to the Land of Wonders,5000,1338,1338,2005,Action and Adventure|Art House and Internation...,15.0,0.866667,-3662,-3662,-0.732400,1.0,7,Wednesday
5938,81,A Plague So Pleasant,1400,0,0,2015,NaN,NaN,NaN,-1400,-1400,-1.000000,NaN,9,Tuesday
5939,82,My Date With Drew,1100,181041,181041,2005,Comedy|Mystery and Suspense|Science Fiction an...,102.0,0.852941,179941,179941,163.582727,1.0,8,Friday


### Director and Writer Metrics

In [559]:
review_metrics = rt_reviews_clean_df.groupby('id').agg(
    review_count=('fresh', 'count'),
    freshness_score=('fresh', lambda x: (x == 'fresh').sum() / len(x))
).reset_index()

In [564]:
# Merge review metrics with director and writer
review_metrics_full = pd.merge(rt_combined_df[['id', 'director', 'writer']], review_metrics, on='id', how='left')
review_metrics_full

,id,director,writer,review_count,freshness_score
0,1,William Friedkin,Ernest Tidyman,NaN,NaN
1,3,David Cronenberg,David Cronenberg|Don DeLillo,163.0,0.631902
2,5,Allison Anders,Allison Anders,23.0,0.782609
3,6,Barry Levinson,Paul Attanasio|Michael Crichton,57.0,0.561404
4,7,Rodney Bennett,Giles Cooper,NaN,NaN
...,...,...,...,...,...
1555,1996,Unknown,Unknown,143.0,0.671329
1556,1997,Steve Barron,Terry Turner|Tom Davis|Dan Aykroyd|Bonnie Turner,28.0,0.357143
1557,1998,Gordon Douglas,Unknown,2.0,1.000000
1558,1999,David Mickey Evans,David Mickey Evans|Robert Gunter,46.0,0.586957


In [565]:
# Group by director
director_metrics = review_metrics_full.groupby('director').agg(
    director_avg_freshness=('freshness_score', 'mean'),
    director_avg_reviews=('review_count', 'mean'),
    director_movie_count=('id', 'count')
).reset_index()

# Group by writer
writer_metrics = review_metrics_full.groupby('writer').agg(
    writer_avg_freshness=('freshness_score', 'mean'),
    writer_avg_reviews=('review_count', 'mean'),
    writer_movie_count=('id', 'count')
).reset_index()

In [566]:
director_metrics

,director,director_avg_freshness,director_avg_reviews,director_movie_count
0,Abel Ferrara,0.293750,10.5,2
1,Adam Ravetch|Sarah Robertson,0.612903,93.0,1
2,Adam Shankman,0.044118,68.0,1
3,Adam Silver|Ben Hayflick,NaN,NaN,1
4,Aditya Chopra,NaN,NaN,1
...,...,...,...,...
1121,Youssef Delara,0.600000,5.0,1
1122,Yves Simoneau,0.333333,6.0,1
1123,Zak Penn,NaN,NaN,1
1124,Zeresenay Berhane Mehari,0.851852,27.0,1


In [567]:
writer_metrics

,writer,writer_avg_freshness,writer_avg_reviews,writer_movie_count
0,Christina Trevino,NaN,NaN,1
1,Aaron Kandell|Jordan Kandell|David Branson Smi...,NaN,NaN,1
2,Aaron Latham|James Bridges,0.214286,14.0,1
3,Aaron Seltzer|Jason Friedberg,0.028571,35.0,1
4,Abby Kohn|Marc Silverstein|Dana Fox,NaN,NaN,1
...,...,...,...,...
1065,Yoav Shamir,0.842105,19.0,1
1066,Yoji Yamada|Ichiro Yamamoto|Emiko Hiramatsu,0.833333,18.0,1
1067,Zak Penn|Ernest Cline,0.618421,76.0,1
1068,Zenzo Matsumura,NaN,NaN,1


In [568]:
# top directors avg reviews
top_directors_by_reviews = director_metrics.sort_values(by='director_avg_reviews', ascending=False).head(10)
top_directors_by_reviews

,director,director_avg_freshness,director_avg_reviews,director_movie_count
659,Martin McDonagh,0.911243,338.0,1
625,Luca Guadagnino,0.949091,275.0,1
709,Michel Hazanavicius,0.961538,260.0,1
493,Joel Coen|Ethan Coen,0.926923,260.0,1
76,Ava DuVernay,0.988372,258.0,1
727,Morten Tyldum,0.896414,251.0,1
678,Mel Gibson,0.485944,249.0,1
354,Guillermo del Toro,0.717300,237.0,1
1002,Steven Soderbergh,0.926724,232.0,1
935,Sam Mendes,0.795158,218.5,2


In [570]:
# top writers avg reviews
top_writers_by_reviews = writer_metrics.sort_values(by='writer_avg_reviews', ascending=False).head(10)
top_writers_by_reviews


,writer,writer_avg_freshness,writer_avg_reviews,writer_movie_count
632,Martin McDonagh,0.911243,338.0,1
423,James Ivory,0.949091,275.0,1
499,John Logan|Robert Wade|Neal Purvis,0.916031,262.0,1
672,Michel Hazanavicius,0.961538,260.0,1
302,Ethan Coen|Joel Coen,0.926923,260.0,1
759,Paul Webb,0.988372,258.0,1
789,Phyllis Nagy,0.941860,258.0,1
359,Graham Moore,0.896414,251.0,1
650,Mel Gibson|Benedict Fitzgerald,0.485944,249.0,1
770,Peter Jackson|Philippa Boyens|Guillermo del To...,0.621399,243.0,1
